pkg/lock/keylock.go

```diff
@@ -0,0 +1,272 @@
+func (l *KeyLock) TrySharedKeyLock() error {
+	return l.lock(keyLockShared|keyLockNonBlocking, defaultLockRetries)
+}
+
+// TrySharedLock takes a co-operative (shared) lock on a key without blocking.
+// lockDir is the directory where the lock file will be created.
+// It will return ErrLocked if an exclusive lock already exists on the key.
+func TrySharedKeyLock(lockDir string, key string) (*KeyLock, error) {
+	return createAndLock(lockDir, key, keyLockShared|keyLockNonBlocking)
+}
+
+// SharedLock takes a co-operative (shared) lock on a key.
+// This is idempotent when the KeyLock already represents a shared lock,
+// and demotes an exclusive lock to shared atomically.
+// It will block if an exclusive lock is already held on the key.
+func (l *KeyLock) SharedKeyLock() error {
+	return l.lock(keyLockShared, defaultLockRetries)
+}
+
+// SharedLock takes a co-operative (shared) lock on a key.
+// lockDir is the directory where the lock file will be created.
+// It will block if an exclusive lock is already held on the key.
+func SharedKeyLock(lockDir string, key string) (*KeyLock, error) {
+	return createAndLock(lockDir, key, keyLockShared)
+}
+
+func createAndLock(lockDir string, key string, mode keyLockMode) (*KeyLock, error) {
+	keyLock, err := NewKeyLock(lockDir, key)
+	if err != nil {
+		return nil, err
+	}
+	err = keyLock.lock(mode, defaultLockRetries)
+	if err != nil {
+		return nil, err
+	}
+	return keyLock, nil
+}
+
+// lock is the base function to take a lock and handle changed lock files
+// As there's the need to remove unused (see CleanKeyLocks) lock files without
+// races, a changed file detection is needed.
+//
+// Without changed file detection this can happen:
+//
+// Process A takes exclusive lock on file01
+// Process B waits for exclusive lock on file01.
+// Process A deletes file01 and then releases the lock.
+// Process B takes the lock on the removed file01 as it has the fd opened
+// Process C comes, creates the file as it doesn't exists, and it also takes an exclusive lock.
+// Now B and C thinks to own an exclusive lock.
+//
+// maxRetries can be passed, useful for testing.
+func (l *KeyLock) lock(mode keyLockMode, maxRetries int) error {
+	retries := 0
+	for {
+		var err error
+		var isExclusive bool
+		var isNonBlocking bool
+		if mode&keyLockExclusive != 0 {
+			isExclusive = true
+		}
+		if mode&keyLockNonBlocking != 0 {
+			isNonBlocking = true
+		}
+		switch {
+		case isExclusive && !isNonBlocking:
+			err = l.keyLock.ExclusiveLock()
+		case isExclusive && isNonBlocking:
+			err = l.keyLock.TryExclusiveLock()
+		case !isExclusive && !isNonBlocking:
+			err = l.keyLock.SharedLock()
+		case !isExclusive && isNonBlocking:
+			err = l.keyLock.TrySharedLock()
+		}
+		if err != nil {
+			return err
+		}
+
+		// Check that the file referenced by the lock fd is the same as
+		// the current file on the filesystem
+		var lockStat, curStat syscall.Stat_t
+		lfd, err := l.keyLock.Fd()
+		if err != nil {
+			return err
```

Comment:
At this point keylock is locked. Shouldn't we unlock or close the keylock on error?